internal/controlplane/handlers_githubwebhooks.go

```diff
@@ -318,15 +317,6 @@ func (s *Server) parseGithubEventForProcessing(
 		return fmt.Errorf("error getting provider: %w", err)
 	}
 
-	pbOpts := []providers.ProviderBuilderOption{
-		providers.WithProviderMetrics(s.provMt),
-		providers.WithRestClientCache(s.restClientCache),
-	}
-	provBuilder, err := providers.GetProviderBuilder(ctx, prov, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
-	if err != nil {
-		return fmt.Errorf("error building client: %w", err)
-	}
```

Comment:
I like removing this factory from all the calls, but I feel like I want two layers of objects:

1. A `ProviderFactory` which takes all the current arguments except `ctx` and `prov`, and has `Build(ctx, currentProject, providerName) (Provider, error)`.  This combines lines 313-328 into a function, and doesn't directly expose `db.Provider` elsewhere.  As we discover other cases where we need to find a provider, the `ProviderFactory` would get new methods, such as `BuildForEntity(ctx, currentProject, entityType)`

3. `fooProvider` objects which wrap the `db.Provider` type with the selected implementation behavior, e.g. `ghAppProvider` and `ghOAuthProvider`.  Both of these would implement `v1.Provider`, which would have `As$Interface` (`AsGit`, `AsGitHub`, `AsREST`, etc) methods for down-casting as well as `Implements` if we need the checks.

I'm a little concerned with passing db row objects everywhere in that we end up making it hard to separate the schema from the business logic; ideally the db queries and token / crypto pieces would live in `ghAppProvider` and `ghOAuthProvider` and we have a central place to implement the mapping from user-provider context -> provider, rather than going directly to the database layer.

Reply:
I completely agree with you, but I also feel like it's a good idea to incrementally move towards that sort of model. Right now, too much knowledge of the creation of profiles exist in too many parts of the codebase. I want to reduce that coupling/lack of cohesion before introducing a new abstraction.

Reply:
I'm fine with incremental progress, as long as we agree on the destination.  I've just had bad experiences with incremental progress hiding disagreement on the final destination.